packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js

```diff
@@ -1,61 +1,124 @@
+    const { supportedFulfillmentTypes } = item;
+
+    // This is a new optional field that UI can pass in case the user selects fulfillment type
+    // for each item in the product details page instead of waiting till checkout
+    let { selectedFulfillmentType } = item;
+
     // Do not re-allocate the item if it is already in the group. Otherwise difficult for other code
     // to create and manage fulfillment groups
-    const itemAlreadyInTheGroup = currentGroups.find(({ itemIds }) => itemIds && itemIds.includes(item._id));
-    if (itemAlreadyInTheGroup) return;
+    // Commenting out the below check since the item by default (if no fulfillment type is provided) will be assigned to 'undecided' group
+    // and later 'setFulfillmentTypeForItems' mutation can re-assign the group by passing an updated 'selectedFulfillmentType'.
+    // const itemAlreadyInTheGroup = currentGroups.find(({ itemIds }) => itemIds && itemIds.includes(item._id));
+    // if (itemAlreadyInTheGroup) return;
 
     if (!supportedFulfillmentTypes || supportedFulfillmentTypes.length === 0) {
-      supportedFulfillmentTypes = ["shipping"];
+      // supportedFulfillmentTypes = ["shipping"];
+      // we no longer default to shipping as fulfillment type
+      throw new ReactionError("not-found", "Product does not have any supported FulfillmentTypes");
+    }
+
+    if (selectedFulfillmentType && !supportedFulfillmentTypes.includes(selectedFulfillmentType)) {
+      throw new ReactionError("not-found", "Selected fulfillmentType is not supported by the Product");
     }
 
     // Out of the current groups, returns the one that this item should be in by default, if it isn't
     // already in a group
-    const group = determineInitialGroupForItem(currentGroups, supportedFulfillmentTypes, item.shopId);
+    // If selectedFulfillmentType is provided, move the item to that group, else add item to undecided group
+    if (!selectedFulfillmentType) selectedFulfillmentType = "undecided";
 
-    if (!group) {
-      // If no compatible group, add one with initially just this item in it
-      currentGroups.push({
-        _id: Random.id(),
-        itemIds: [item._id],
-        shopId: item.shopId,
-        type: supportedFulfillmentTypes[0]
-      });
-    } else if (!group.itemIds) {
-      // If there is a compatible group but it has no items array, add one with just this item in it
-      group.itemIds = [item._id];
-    } else if (!group.itemIds.includes(item._id)) {
-      // If there is a compatible group with an items array but it is missing this item, add this item ID to the array
-      group.itemIds.push(item._id);
-    }
-  });
+    // check if the selectedFulfillmentType is an 'enabled' fulfillmentType, if not set is 'undecided'
+    /* eslint-disable no-await-in-loop */
+    const enabledFulfillmentTypeObjs = await Fulfillment.find({ "shopId": item.shopId, "provider.enabled": true }).toArray();
+    /* eslint-enable no-await-in-loop */
```

Comment:
And this one?

Reply:
Linter thorws error
`Unexpected await inside a loop.eslint[no-await-in-loop](https://eslint.org/docs/rules/no-await-in-loop)`